src/guardian/duplicates/keys-duplication-checker.service.ts

```diff
@@ -77,35 +51,107 @@ export class KeysDuplicationCheckerService {
     return { duplicates, unresolved };
   }
 
-  public findDuplicateKeys(keys: RegistryKey[]): [string, RegistryKey[]][] {
-    const keyOccurrencesMap = keys.reduce((acc, key) => {
-      const occurrences = acc.get(key.key) || [];
-      occurrences.push(key);
-      acc.set(key.key, occurrences);
+  /**
+   * Groups keys by their pubkey and returns a list of those with duplicates.
+   *
+   * This method iterates over the provided keys and groups them by their unique pubkey.
+   * It then filters out any groups that do not have duplicates, returning only the groups
+   * that contain more than one instance of the pubkey.
+   *
+   * @param keys - An array of `RegistryKey` objects to be checked for duplicates.
+   * @returns An array of tuples where each tuple contains a pubkey string and an array of
+   *          `RegistryKey` objects that share that pubkey. Only keys with duplicates are included.
+   */
+  public getDuplicateKeyGroups(keys: RegistryKey[]): [string, RegistryKey[]][] {
+    const keyMap = keys.reduce((acc, key) => {
+      const duplicateKeys = acc.get(key.key) || [];
+      duplicateKeys.push(key);
+      acc.set(key.key, duplicateKeys);
 
       return acc;
     }, new Map<string, RegistryKey[]>());
 
-    return Array.from(keyOccurrencesMap.entries()).filter(
-      ([, occurrences]) => occurrences.length > 1,
+    return Array.from(keyMap.entries()).filter(
+      ([, duplicateKeys]) => duplicateKeys.length > 1,
+    );
+  }
+
+  private async processDuplicateKeyGroup(
+    key: string,
+    suspectedDuplicateKeys: RegistryKey[],
+    blockData: BlockData,
+  ): Promise<{ duplicates: RegistryKey[]; unresolved: RegistryKey[] }> {
+    const operators = this.getOperators(suspectedDuplicateKeys);
+
+    if (operators.length === 1) {
+      return this.handleSingleOperatorDuplicates(suspectedDuplicateKeys);
+    }
+
+    if (this.hasDepositedKey(suspectedDuplicateKeys)) {
+      return this.handleDepositedKeyDuplicates(suspectedDuplicateKeys);
+    }
+
+    return await this.handleMultiOperatorDuplicates(
+      key,
+      suspectedDuplicateKeys,
+      operators,
+      blockData,
     );
   }
 
-  private extractOperators(occurrences: RegistryKey[]): Set<string> {
-    return new Set(
-      occurrences.map((key) => `${key.moduleAddress}-${key.operatorIndex}`),
+  private getOperators(keys: RegistryKey[]): string[] {
```

Comment:
method does not accurately reflect the essence: this method returns idempotency keys, I would suggest to reflect this in the name and in all arguments of functions to which it is passed.